groupnorm: graph mode static quant support

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39413 graph mode: util for fusion of functions which require observation
* #39203 quant layer/group/instance norm: make weights and biases optional
* #39096 instancenorm: static quant graph mode support
* **#39095 groupnorm: graph mode static quant support**
* #39094 layernorm: eager mode qat support
* #39093 instancenorm: eager mode QAT support
* #39092 groupnorm: eager mode QAT support
* #39091 instancenorm: eager static quant support
* #39090 groupnorm: eager static quant support
* #37802 quant: clean up normalization channels_last handling

Summary:

Hooks up groupnorm to graph mode static quant

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885257](https://our.internmc.facebook.com/intern/diff/D21885257)

vkuzo committed:
groupnorm: graph mode static quant support

Summary:

Hooks up groupnorm to graph mode static quant

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization/quantization_patterns.h:
```diff
@@ -617,6 +617,19 @@ graph(%a_quant, %normalized_shape, %weight, %bias, %eps, %cudnn_enabled, %output
          return (%r) )";
 
+  // quantized::group_norm
+  std::string group_norm = R"(
+graph(%a_quant, %num_groups, %weight, %bias, %eps, %cudnn_enabled, %output_scale, %output_zero_point, %scalar_type):
+         %a_dequant = aten::dequantize(%a_quant)
+         %r_gn = aten::group_norm(%a_dequant, %num_groups, %weight, %bias, %eps, %cudnn_enabled)
+         %r = aten::quantize_per_tensor(%r_gn, %output_scale, %output_zero_point, %scalar_type)
+         return (%r) )";
+
+  std::string quantized_group_norm = R"(
+graph(%a_quant, %num_groups, %weight, %bias, %eps, %cudnn_enabled, %output_scale, %output_zero_point, %scalar_type):
+         %r = quantized::group_norm(%a_quant, %num_groups, %weight, %bias, %eps, %output_scale, %output_zero_point)
+         return (%r) )";
```
I was thinking of refactoring this into a function actually, so that adding new pattern would require minimal code

jerryzh168 reviewed (approved):
please refactor the code to have a function that can generate the patterns if you have time, example: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/quantization/quantization_patterns.h#L73

vkuzo commented:
> please refactor the code to have a function that can generate the patterns if you have time, example: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/quantization/quantization_patterns.h#L73

sure thing.  Let me do that in a follow-up PR.

vkuzo committed:
Update on "groupnorm: graph mode static quant support"


Summary:

Hooks up groupnorm to graph mode static quant

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "groupnorm: graph mode static quant support"


Summary:

Hooks up groupnorm to graph mode static quant

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "groupnorm: graph mode static quant support"


Summary:

Hooks up groupnorm to graph mode static quant

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885257](https://our.internmc.facebook.com/intern/diff/D21885257)

[ghstack-poisoned]

vkuzo committed:
Update on "groupnorm: graph mode static quant support"


Summary:

Hooks up groupnorm to graph mode static quant

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885257](https://our.internmc.facebook.com/intern/diff/D21885257)

[ghstack-poisoned]

vkuzo committed:
Update on "groupnorm: graph mode static quant support"


Summary:

Hooks up groupnorm to graph mode static quant

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885257](https://our.internmc.facebook.com/intern/diff/D21885257)

[ghstack-poisoned]

vkuzo committed:
Update on "groupnorm: graph mode static quant support"


Summary:

Hooks up groupnorm to graph mode static quant

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885257](https://our.internmc.facebook.com/intern/diff/D21885257)

[ghstack-poisoned]
